sdk/pinocchio/src/sysvars/tests/slot_hashes_tests.rs

```diff
@@ -0,0 +1,855 @@
+        let slot_hashes_wrong2 = unsafe { SlotHashes::new_unchecked(data.as_slice(), 3) };
+        assert!(slot_hashes_wrong2.get_entry_count().is_err());
+
+        let slot_hashes_consistent = unsafe { SlotHashes::new_unchecked(data.as_slice(), 2) };
+        assert_eq!(slot_hashes_consistent.get_entry_count().unwrap(), 2);
+    }
+
+    #[test]
+    fn test_entries_exposed_no_std() {
+        let entries = generate_mock_entries(8, 80, DecrementStrategy::Strictly1);
+        let data = create_mock_data(&entries);
+        let sh = unsafe { SlotHashes::new_unchecked(data.as_slice(), entries.len()) };
+
+        let slice = sh.entries();
+        assert_eq!(slice.len(), entries.len());
+        for (i, e) in slice.iter().enumerate() {
+            assert_eq!(e.slot(), entries[i].0);
+            assert_eq!(e.hash, entries[i].1);
+        }
+    }
+
+    #[test]
+    fn test_fetch_into_offset_validation() {
+        let buffer_len = 200;
+
+        // Offset 0 (start of data) - should pass validation
+        assert!(SlotHashes::<&[u8]>::validate_fetch_offset(0, buffer_len).is_ok());
+
+        // Offset 8 (start of first entry) - should pass validation
+        assert!(SlotHashes::<&[u8]>::validate_fetch_offset(8, buffer_len).is_ok());
+
+        // Offset 48 (start of second entry) - should pass validation
+        assert!(SlotHashes::<&[u8]>::validate_fetch_offset(48, buffer_len).is_ok());
+
+        // Offset 88 (start of third entry) - should pass validation
+        assert!(SlotHashes::<&[u8]>::validate_fetch_offset(88, buffer_len).is_ok());
+
+        // Invalid offsets that should fail validation
+
+        // Offset beyond MAX_SIZE
+        assert!(SlotHashes::<&[u8]>::validate_fetch_offset(MAX_SIZE as u64, buffer_len).is_err());
+
+        // Offset pointing mid-entry (not aligned)
+        assert!(SlotHashes::<&[u8]>::validate_fetch_offset(12, buffer_len).is_err()); // 8 + 4, mid-entry
+        assert!(SlotHashes::<&[u8]>::validate_fetch_offset(20, buffer_len).is_err()); // 8 + 12, mid-entry
+        assert!(SlotHashes::<&[u8]>::validate_fetch_offset(35, buffer_len).is_err()); // 8 + 27, mid-entry
+
+        // Offset in header but not at start
+        assert!(SlotHashes::<&[u8]>::validate_fetch_offset(4, buffer_len).is_err()); // Mid-header
+        assert!(SlotHashes::<&[u8]>::validate_fetch_offset(7, buffer_len).is_err()); // End of header
+
+        // Test buffer + offset exceeding MAX_SIZE
+        assert!(SlotHashes::<&[u8]>::validate_fetch_offset(1, MAX_SIZE).is_err());
+        assert!(SlotHashes::<&[u8]>::validate_fetch_offset(MAX_SIZE as u64 - 100, 200).is_err());
+
+        // Last entry
+        assert!(
+            SlotHashes::<&[u8]>::validate_fetch_offset(8 + 511 * ENTRY_SIZE as u64, 40).is_ok()
+        );
+
+        // One past last valid entry
+        assert!(
+            SlotHashes::<&[u8]>::validate_fetch_offset(8 + 512 * ENTRY_SIZE as u64, 40).is_err()
+        );
+    }
+}
+
+#[cfg(test)]
+mod edge_tests {
```

Comment:
*bikeshedding*

You can probably put different tests into different modules as well.
`slot_hashes/tests.rs`, `slot_hashes/edge_tests.rs` and so on.